compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -29,6 +29,64 @@ pub(super) fn opt_const_param_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<
         let parent_node = tcx.hir().get(parent_node_id);
 
         match parent_node {
+            // This matches on types who's paths couldn't be resolved without typeck'ing e.g.
+            //
+            // trait Foo {
+            //   type Assoc<const N1: usize>;
+            //   fn foo() -> Self::Assoc<3>;
+            //   // note: if the def_id argument is the 3 then in this example
+            //   // parent_node would be the node for Self::Assoc<_>
+            // }
+            // We didnt write <Self as Foo>::Assoc so the Self::Assoc<_> is lowered to QPath::TypeRelative.
+            // I believe this match arm is only needed for GAT but I am not 100% sure - BoxyUwU
+            Node::Ty(hir_ty @ Ty { kind: TyKind::Path(QPath::TypeRelative(_, segment)), .. }) => {
+                // Walk up from the parent_node to find an item so that
+                // we can resolve the relative path to an actual associated type.
+                // For the code example above, this item would be the Foo trait.
+                let item_hir_id = tcx
+                    .hir()
+                    .parent_iter(hir_id)
+                    .filter(|(_, node)| matches!(node, Node::Item(_)))
+                    .map(|(id, _)| id)
+                    .next()
+                    .unwrap();
+                let item_did = tcx.hir().local_def_id(item_hir_id).to_def_id();
+                let item_ctxt = &ItemCtxt::new(tcx, item_did) as &dyn crate::astconv::AstConv<'_>;
+
+                // This ty will be the actual associated type so that we can
+                // go through its generics to find which param our def_id corresponds to.
+                // For the code example above, this ty would be the Assoc<const N1: usize>.
+                let ty = item_ctxt.ast_ty_to_ty(hir_ty);
+                if let ty::Projection(projection) = ty.kind() {
+                    let generics = tcx.generics_of(projection.item_def_id);
+
+                    let arg_index = segment
```

Comment:
"Find the index where the `3` argument appears in the projection"

Reply:
Side note that I don't know why we are filtering to `is_const`

Reply:
It's not bad, just seems unnecessary

Reply:
I asked lcnr why we do this and it doesn't seem strictly necessary but was put there in case we're ever allowed to elide lifetimes in the segment. That isn't allowed atm though so the filter is more of a precaution ^^

Reply:
I see. OK. I would never want to allow that, but it's ok.